File: 518/coin-change-ii.c
int change(int amount, int* coins, int coinsSize){
    int table[amount+1];
    for(int i=0; i<=amount; i++){ //第0種幣 coins[0] 
        if(i%coins[0]==0) table[i]=1; //是倍數時，便能用此幣支付
        else table[i]=0; //不是的話，就沒有用此幣的可能
    }//先建出基礎的表格
    //接著測試其他的幣
    for(int c=1; c<coinsSize; c++){
        int c2 = coins[c];
        for(int i=c2; i<=amount; i++){
        //i不夠大時，沒辦法使用 c2 這種幣
            table[i] += table[i-c2]; //夠大時，能逐一使用此幣，增加支付的可能性
        }
    }
    return table[amount];
}

File: 518/coin-change-ii.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        combN = [0]*(amount+1)
        # combN[i] 表示合計i元，有多少種組合
        combN[0] = 1 # 完全不用任何coin的這種方法，就可得到0元
        for c in coins:
            for i in range(c, amount+1):
                combN[i] += combN[i-c] # 增加可能的組合
        
        return combN[amount]
        
